applications: nrf_desktop: Forward boot report doc (#5590)
nrfconnect/sdk-nrf: applications/nrf_desktop/doc/hid_forward.rst

Use the |hid_forward| to:

* Receive the HID input reports from the peripherals connected over Bluetooth®.

> * Can process the HID input reports in report or boot protocol.

Reviewer: You have a grammar issue here:
"Use the |hid_forward| to: * Can process the HID input reports in report or boot protocol."
I guess this would do?
```suggestion
* Process the HID input reports in report or boot protocol.
```

> #. The |hid_forward| is enabled with the :kconfig:`CONFIG_DESKTOP_HID_FORWARD_ENABLE` option.
>    This option is available only if you enable both the :kconfig:`CONFIG_BT_CENTRAL` and :kconfig:`CONFIG_BT_HOGP` Kconfig options.

Reviewer: ```suggestion
#. Make sure to enable both the :kconfig:`CONFIG_BT_CENTRAL` and :kconfig:`CONFIG_BT_HOGP` Kconfig options. 
#. Enable |hid_forward| using the :kconfig:`CONFIG_DESKTOP_HID_FORWARD_ENABLE` option.
   This option is available only if you enable both options from the previous step.
```

   The module is used on Bluetooth Central to forward the HID reports that are received by the HID service client.

> #. To forward keyboard boot reports enable option: :kconfig:`CONFIG_DESKTOP_HID_BOOT_INTERFACE_KEYBOARD`.
>    To forward mouse boot reports enable option: :kconfig:`CONFIG_DESKTOP_HID_BOOT_INTERFACE_MOUSE`.
>    Dongle forwards HID reports from both mouse and keyboard and therefore it may act either as boot mouse or boot keyboard.
>    Those options affect :ref:`nrf_desktop_usb_state` that subscribes for HID boot keyboard reports.
>    For more information about configuration of HID boot protocol see Boot protocol configuration section in :ref:`nrf_desktop_usb_state`

Reviewer: ```suggestion
#. Enable one of the following options:

   * :kconfig:`CONFIG_DESKTOP_HID_BOOT_INTERFACE_KEYBOARD` - This option enables forwarding keyboard boot reports.
   * :kconfig:`CONFIG_DESKTOP_HID_BOOT_INTERFACE_MOUSE` - This option enables forwarding mouse boot reports.
   
   Those options affect :ref:`nrf_desktop_usb_state` that subscribes for HID boot keyboard reports.
   The Dongle forwards HID reports from both mouse and keyboard, and so either option works if you want to have the Dongle work as boot mouse or boot keyboard.
   For more information about the configuration of the HID boot protocol, see the boot protocol configuration section in the :ref:`nrf_desktop_usb_state` documentation.
   
```

#. The |hid_forward| is enabled with the :kconfig:`CONFIG_DESKTOP_HID_FORWARD_ENABLE` option.
   This option is available only if you enable both the :kconfig:`CONFIG_BT_CENTRAL` and :kconfig:`CONFIG_BT_HOGP` Kconfig options.
   The module is used on Bluetooth Central to forward the HID reports that are received by the HID service client.

> #. To enable forwarding of boot protocol reports enable options: :kconfig:`CONFIG_USB_HID_BOOT_PROTOCOL`, :kconfig:`CONFIG_DESKTOP_HID_BOOT_INTERFACE_KEYBOARD`

Reviewer 2: I would prefer do avoid adding references to Kconfig options related to USB in this module (apart from CONFIG_USB_HID_DEVICE_COUNT that affects the module directly)

Reviewer 2: I think we should either:
- refer to USB state module that already describes relevant USB-related options; BOOT_INTERFACE options can be described in hid_state (peripherals) and hid_forward (centrals)
- Introduce separated, more general section relate to HID boot protocol configuration in main doc page (if you think this is actually needed)

Reviewer 2: btw. boot protocol code can be either 1 or 2, which means respectively boot keyboard or boot mouse

#. The |hid_forward| is enabled with the :kconfig:`CONFIG_DESKTOP_HID_FORWARD_ENABLE` option.
   This option is available only if you enable both the :kconfig:`CONFIG_BT_CENTRAL` and :kconfig:`CONFIG_BT_HOGP` Kconfig options.
   The module is used on Bluetooth Central to forward the HID reports that are received by the HID service client.
#. To forward keyboard boot reports enable option: :kconfig:`CONFIG_DESKTOP_HID_BOOT_INTERFACE_KEYBOARD`.

>    This option affects usb_state module that subscribes for HID boot keyboard reports.

Reviewer 2: :ref:`nrf_desktop_usb_state`

When reports are forwarded to the :ref:`nrf_desktop_usb_state`, each HID-class USB device is treated independently.
This means that the |hid_forward| can concurrently send reports to every available HID-class USB device, if a report from the linked peripheral is available.

> HID boot protocol reports are received with report ID equal :c:enum:`REPORT_ID_RESERVED` and it is later changed based on  :c:struct:`bt_hogp` objects by |hid_forward| to either :c:enum:`REPORT_ID_BOOT_KEYBOARD` or :c:enum:`REPORT_ID_BOOT_MOUSE`.

Reviewer: ```suggestion
When received by the module, the HID boot protocol reports come with the report ID assigned to :c:enum:`REPORT_ID_RESERVED`.
This ID is later changed by |hid_forward| to either :c:enum:`REPORT_ID_BOOT_KEYBOARD` or :c:enum:`REPORT_ID_BOOT_MOUSE`.
The choice depends on :c:struct:`bt_hogp` objects.
```

> While receiving subscription event, |hid_forward| updates HID protocol of Bluetooth connected peripherals that are associated with given subscriber.

Reviewer: ```suggestion
While receiving a subscription event, the |hid_forward| updates the HID protocol of peripherals connected through Bluetooth that are associated with the given subscriber.
```

Forwarding HID input reports
============================

After :ref:`nrf_desktop_ble_discovery` successfully discovers a connected peripheral, the |hid_forward| automatically subscribes for every HID input report provided by the peripheral.

> Subscriber can use either HID boot protocol or HID report protocol, both protocols cannot be used at the same time.

Reviewer: ```suggestion
The subscriber can use either the HID boot protocol or the HID report protocol, but both protocols cannot be used at the same time.
```

> With current implementation, the :ref:`nrf_desktop_usb_state` supports either HID boot keyboard or HID boot mouse reports, because the HID boot protocol code is set using a single Kconfig option that is common for all of the instances of USB HID.

Reviewer: ```suggestion
In the current implementation, the :ref:`nrf_desktop_usb_state` supports either the HID boot keyboard or the HID boot mouse reports, because the HID boot protocol code is set using a single Kconfig option that is common for all of the instances of the USB HID.
```

============================

After :ref:`nrf_desktop_ble_discovery` successfully discovers a connected peripheral, the |hid_forward| automatically subscribes for every HID input report provided by the peripheral.

> Reports in both HID report protocol and HID boot protocol can be used by the subscriber.

Reviewer 2: Subscriber can use either HID boot or HID report protocol, both protocols cannot be used at the same time.

> However with current implementation, the USB state module supports either HID boot keyboard or HID boot mouse reports, because the boot report type is set as a single Kconfig option.

Reviewer 2: :ref:`nrf_desktop_usb_state`

The :c:func:`hidc_read` callback is called when HID input report is received from the connected peripheral.
The received HID input report data is converted to ``hid_report_event``.

``hid_report_event`` is submitted and then the HID-class USB device configured by :ref:`nrf_desktop_usb_state` forwards it to the host.
When a HID report is sent to the host by the HID-class USB device, the |hid_forward| receives a ``hid_report_sent_event`` with the identifier of this device.

> HID boot protocol reports are received with report ID equal REPORT_ID_RESERVED and it is later changed based on hogp objects by |hid_forward| to either REPORT_ID_BOOT_KEYBOARD or REPORT_ID_BOOT_MOUSE.

Reviewer 2: :c:enum:`REPORT_ID_RESERVED`

Reviewer 2: :c:enum:`REPORT_ID_BOOT_KEYBOARD`

> While receiving subscription event, |hid_forward| updates report protocol in case if report protocol of the subscriber had changed.

Reviewer 2: I would move it to interaction with USB, since we describe subscription mechanism there

When reports are forwarded to the :ref:`nrf_desktop_usb_state`, each HID-class USB device is treated independently.
This means that the |hid_forward| can concurrently send reports to every available HID-class USB device, if a report from the linked peripheral is available.

When received by the module, the HID boot protocol reports come with the report ID assigned to :c:enum:`REPORT_ID_RESERVED`.
This ID is later changed by |hid_forward| to either :c:enum:`REPORT_ID_BOOT_KEYBOARD` or :c:enum:`REPORT_ID_BOOT_MOUSE`.

> The choice depends on :c:struct:`bt_hogp` objects.

Reviewer 2: depends on fact, if the Bluetooth device is boot mouse or keyboard ? I am not sure if we should refer to internal data structures here.

On a peripheral disconnection, nRF Desktop central informs the host that all the pressed keys reported by the peripheral are released.
This is done to make sure that user will not observe a problem with a key stuck on peripheral disconnection.

> This feature is not yet implemented for boot protocole reports.

Reviewer 2: This is not true. Actually FW would send empty reports in HID report protocol no matter which protocol is selected.
I think we need https://github.com/nrfconnect/sdk-nrf/pull/5584 before the release (otherwise we must add it to known issues).
